ocaml/xapi/cli_operations.ml

```diff
@@ -646,27 +646,39 @@ let make_param_funs getall getallrecs getbyuuid record class_name def_filters de
   let p_set (printer : printer) rpc session_id params =
     let record = get_record rpc session_id (List.assoc "uuid" params) in
     let set_params = List.filter (fun (p,_) -> not (List.mem p ("uuid"::stdparams))) params in
+    (* Hashtable set_map_table contains key as set_map function
+       and associated value as list of (key, value) pairs to set a map field *)
+    let set_map_table: (((string * string) list -> unit), (string * string) list) Hashtbl.t = Hashtbl.create 10 in
 
     let set_field (k,v) =
       let field_type = get_field_type k record in
       match field_type with
       | Map s ->
-        let field=field_lookup record s in
+        let field = field_lookup record s in
         let n = String.length s in
         let key = String.sub k (n + 1) (String.length k - n - 1) in
         let get_map = match field.get_map with
           | Some x -> x
           | None -> failwith (Printf.sprintf "Broken Records (field %s)" s)
         in begin
           (* If set_in_map is present, use it instead of using remove_from_map followed by add_to_map. *)
-          match field.set_in_map with
-          | Some set_in_map -> set_in_map key v
-          | None ->
-            let add_to_map = match field.add_to_map with Some f -> f | None -> failwith ("Map field '"^s^"' is read-only.") in
-            let remove_from_map = match field.remove_from_map with Some f -> f | None -> failwith (Printf.sprintf "Records broken (field %s)" s) in
-            let map = get_map () in
-            if List.mem_assoc key map then remove_from_map key;
-            add_to_map key v
+          (* If set_map is present then accumulate all (key, value) pairs into set_map_table *)
+          match field.set_in_map, field.set_map  with
+          | Some set_in_map, None -> set_in_map key v
+          | None, Some set_map ->
+            if not (Hashtbl.mem set_map_table set_map) then
```

Comment:
What I meant in my previous comment, was the this whole conditional block can be replaced by just this (`Hashtbl.replace` just adds the k/v pair if the key is not in the table):
```
let existing_params = try Hashtbl.find set_map_table set_map with Not_found -> [] in
Hashtbl.replace set_map_table set_map ((key, v) :: existing_params)
```

Reply:
Updated, Thanks.